### ISSUE TITLE
Automatic differentiation for circuits

### ISSUE THREAD

**Issue by GiggleLiu** [QuantumBFS/Yao.jl#141]
Current differentiation strategy is limited to loyal simulation of quantum differentiation in
https://arxiv.org/abs/1803.00745
The complexity is O(M^2) with M the number of parameters.

Back propagation is much more efficient in simulation, it has O(N) simulation complexity. 

### Example: Classical back propagation
```julia console
julia> circuit = chain(4, repeat(4, H, 1:4), put(4, 3=>Rz(0.5)), control(2, 1=>X), put(4, 4=>Ry(0.2))) |> autodiff(:BP)
Total: 4, DataType: Complex{Float64}
chain
├─ repeat on (1, 2, 3, 4)
│  └─ H gate
├─ [∂] put on (3)
│  └─ Rot Z gate: 0.5
├─ control(2)
│  └─ (1,)=>X gate
└─ [∂] put on (4)
   └─ Rot Y gate: 0.2

julia> op = put(4, 3=>Y);  # loss is defined as its expectation.
julia> ψ = rand_state(4);
julia> ψ |> circuit;
julia> δ = ψ |> op;     # ∂f/∂ψ*
julia> backward(circuit, δ);    # classical back propagation!

julia> g1 = gradient(circuit)  # collect gradient
2-element Array{Float64,1}:
  0.0652344855354505 
 -0.15774423994151457
```

Blocks marked by `∂` sign will be differentiated automatically using `g1 = gradient(circuit, δ)
`, where `δ = ∂f/∂ψ*`.

Then, for the loyal simulation on quantum circuit, we provide `QDiff` interface.
### Example: Practical quantum differenciation
```julia console
julia> c = chain(put(4, 1=>Rx(0.5)), control(4, 1, 2=>Ry(0.5)), kron(4, 2=>Rz(0.3), 3=>Rx(0.7))) |> autodiff(:QC)  # automatically mark differentiable blocks
Total: 4, DataType: Complex{Float64}
chain
├─ put on (1)
│  └─ [̂∂] Rot X gate: 0.5
├─ control(1)
│  └─ (2,)=>Rot Y gate: 0.5
└─ kron
   ├─ 2=>[̂∂] Rot Z gate: 0.3
   └─ 3=>[̂∂] Rot X gate: 0.7

julia> dbs = collect(c, QDiff)  # collect all QDiff blocks
Sequence
├─ [̂∂] Rot X gate: 0.5
├─ [̂∂] Rot Z gate: 0.3
└─ [̂∂] Rot X gate: 0.7

julia> ed = exactdiff(dbs[1].block) do   # the exact differentiation with respect to first QDiff block.
           expect(put(4, 1=>Z), zero_state(4) |> c) |> real
       end
0.47942553860420284 + 0.0im

julia> ed = numdiff(dbs[1].block) do    # compare with numerical differentiation
           expect(put(4, 1=>Z), zero_state(4) |> c) |> real
       end
0.47941754821851923 + 0.0im

julia> loss1z() = expect(kron(4, 1=>Z, 2=>X), copy(reg) |> c) |> real;  # return loss

julia> ed = exactdiff.(loss1z, dbs)   # using broadcast to get all gradients.
3-element Array{Complex{Float64},1}:
     -0.1018526337019543 + 0.0im
    -0.24594877978563112 + 0.0im
 -1.3877787807814457e-17 + 0.0im
```

**Comment by Roger-luo** [comment]
This looks good to me. `gradient` is also compatible with other AD tools. but can you give a single PR of this?

**Comment by GiggleLiu** [comment]
Refactoring is for autodiff.

**Comment by GiggleLiu** [comment]
## Benchmark Result
#### New Diff
```
BenchmarkTools.Trial:
  memory estimate:  277.22 MiB
  allocs estimate:  14966
  --------------
  minimum time:     190.460 ms (52.04% GC)
  median time:      198.205 ms (50.42% GC)
  mean time:        198.573 ms (50.73% GC)
  maximum time:     212.631 ms (48.00% GC)
  --------------
  samples:          26
  evals/sample:     1
```
#### Old Diff
```julia
BenchmarkTools.Trial:
  memory estimate:  8.38 GiB
  allocs estimate:  145223
  --------------
  minimum time:     9.597 s (74.25% GC)
  median time:      9.597 s (74.25% GC)
  mean time:        9.597 s (74.25% GC)
  maximum time:     9.597 s (74.25% GC)
  --------------
  samples:          1
  evals/sample:     1
```

**Comment by GiggleLiu** [comment]
closed by #147

**Comment by GiggleLiu** [comment]
closed by #147
